waku/v2/waku_archive/driver/postgres_driver/asyncpool.nim

```diff
@@ -0,0 +1,202 @@
+
+  if not pool.isLive():
+    return err("pool is not live")
+
+  # stablish new connections if we are under the limit
+  if pool.isBusy() and pool.conns.len < pool.maxConnections:
+    let connRes = connection.open(pool.connString)
+    if connRes.isOk():
+      let conn = connRes.get()
+      pool.conns.add(PgDbConn(dbConn: conn, busy: true, open: true))
+      return ok(pool.conns.len - 1)
+    else:
+      return err("failed to stablish a new connection: " & connRes.error)
+
+  # wait for a free connection without blocking the async runtime
+  while pool.isBusy():
+    await sleepAsync(0.milliseconds)
+
+  for index in 0..<pool.conns.len:
+    if pool.conns[index].busy:
+      continue
+
+    pool.conns[index].busy = true
+    return ok(index)
+
+proc releaseConn(pool: PgAsyncPool, conn: DbConn) =
+  ## Marks the connection as released.
+  for i in 0..<pool.conns.len:
+    if pool.conns[i].dbConn == conn:
+      pool.conns[i].busy = false
+
+proc query*(pool: PgAsyncPool,
+            query: string,
+            args: seq[string] = newSeq[string](0)):
+            Future[Result[seq[Row], string]] {.async.} =
+  ## Runs the SQL query getting results.
+
+  let connIndexRes = await pool.getConnIndex()
+  if connIndexRes.isErr():
+    return err("connRes.isErr in query: " & connIndexRes.error)
+
+  let conn = pool.conns[connIndexRes.value].dbConn
+  defer: pool.releaseConn(conn)
+
+  return await conn.rows(sql(query), args)
+
+proc exec*(pool: PgAsyncPool,
+           query: string,
+           args: seq[string]):
+           Future[ArchiveDriverResult[void]] {.async.} =
+  ## Runs the SQL query without results.
+
+  let connIndexRes = await pool.getConnIndex()
+  if connIndexRes.isErr():
+    return err("connRes is err in exec: " & connIndexRes.error)
+
+  let conn = pool.conns[connIndexRes.value].dbConn
+  defer: pool.releaseConn(conn)
+
+  let rowsRes = await conn.rows(sql(query), args)
+  if rowsRes.isErr():
+    return err("rowsRes is err in exec: " & rowsRes.error)
+
+  return ok()
+
+proc runStmt*(pool: PgAsyncPool,
+              baseStmt: string,
+              args: seq[string]):
+              Future[ArchiveDriverResult[void]] {.async.} =
+  # Runs a stored statement, for performance purposes.
+  # In the current implementation, this is aimed
+  # to run the 'insertRow' stored statement aimed to add a new Waku message.
+
+  let connIndexRes = await pool.getConnIndex()
+  if connIndexRes.isErr():
+    return ArchiveDriverResult[void].err(connIndexRes.error())
+
+  let conn = pool.conns[connIndexRes.value].dbConn
+  defer: pool.releaseConn(conn)
+
+  var preparedStmt = pool.conns[connIndexRes.value].insertStmt
+  if cast[string](preparedStmt) == "":
+    # The connection doesn't have insertStmt set yet. Let's create it.
+    # Each session/connection should have its own prepared statements.
+    try:
+      pool.conns[connIndexRes.value].insertStmt =
+                                    conn.prepare("insertRow", sql(baseStmt), 7)
```

Comment:
I'm probably missing something here, what is 7?

Reply:
7 is the number of parameters to be handled.
In this case, we are only setting an unique "insertRow" prepared statement and the `baseStmt` has the next value:

```
INSERT INTO messages (id, storedAt, contentTopic, payload, pubsubTopic, version, timestamp) VALUES ($1, $2, $3, $4, $5, $6, $7);
```

The concept of "prepared statement" is interesting from a performance point of view.

Reply:
I see! thanks, its much clearer now
Can we make this value a constant near the "prepared statement", in case the value changes in the future?